Initial

henriquejsfj opened the pull request:
All functionalities to make this component work.
Depends on :
- [ ] https://github.com/rock-core/rock-package_set/pull/219

Commit:
Update dependecies to properly build the components and the tests.

Commit:
Add the task.

Commit:
Add the test.

henriquejsfj reviewed (commented):
Self-review done, your turn @doudou

doudou commented on tasks/Task.cpp:
```diff
@@ -0,0 +1,152 @@
+        raw_cmd_obj.buttonValue[6] = jdata["buttons"][8].asDouble() > _button_threshold.get();
+        raw_cmd_obj.buttonValue[7] = jdata["buttons"][9].asDouble() > _button_threshold.get();
+
+    } catch (std::exception e){
```
If you are blanket-catching exceptions like this, at least add a comment to explain in which situations this exception could be thrown, and explain why it's fine for you to ignore it.

doudou commented on tasks/Task.cpp:
```diff
@@ -0,0 +1,152 @@
+        raw_cmd_obj.buttonValue[7] = jdata["buttons"][9].asDouble() > _button_threshold.get();
+
+    } catch (std::exception e){
+        std::cout << e.what() << std::endl;
```
Use `LOG_ERROR_S` from `base-logging`. See [this file](https://github.com/rock-control/control-orogen-motor_controller/blob/fix_corner_cases_and_input_validation/tasks/PIDTask.cpp#L4), and don't forget to add `base/logging` to `manifest.xml` as well as adding the relevant `using_library` to the orogen file.

doudou commented on tasks/Task.cpp:
```diff
@@ -0,0 +1,152 @@
+bool Task::updateRawCommand(const char *data){
+    Json::Value jdata;
+    Json::CharReaderBuilder rbuilder;
+    std::unique_ptr<Json::CharReader> const reader(rbuilder.newCharReader());
```
Do you have to create it at each request ? I'm guessing it's probably relatively expensive to build ...

doudou commented on tasks/Task.hpp:
```diff
@@ -0,0 +1,64 @@
+    std::shared_ptr<JoystickHandler> handler;
+    controldev::RawCommand raw_cmd_obj;
+    bool controlling;
+    bool updateRawCommand(const char *data);
```
Indentation looks completely broken. Using tabs ?

doudou commented on tasks/Task.cpp:
```diff
@@ -0,0 +1,152 @@
+    : TaskBase(name, initial_state)
+{
+    for (int i=1; i<=8; ++i){
+        raw_cmd_obj.buttonValue.push_back(0);
```
`std::vector`'s resize allows to initialize elements, `resize(8, 0)` would be fine here.

doudou commented on tasks/Task.cpp:
```diff
@@ -0,0 +1,152 @@
+
+    thread = new std::thread([this]{ this->server->loop(); });
+
+    return true;
```
duplicated line

doudou commented on tasks/Task.cpp:
```diff
@@ -0,0 +1,152 @@
+void Task::errorHook()
+{
+    TaskBase::errorHook();
+    server->terminate();
```
`errorHook` is a **runtime** error, i.e. it is meant to go back to normal operations afterwards.

The hook 'on error' is `exceptionHook`, which automatically goes through stopHook and cleanupHook anyways. So just do cleanup in those.

doudou commented on tasks/Task.cpp:
```diff
@@ -0,0 +1,152 @@
+        raw_cmd_obj.buttonValue[4] = jdata["buttons"][4].asDouble() > _button_threshold.get();
+        raw_cmd_obj.buttonValue[5] = jdata["buttons"][5].asDouble() > _button_threshold.get();
+        raw_cmd_obj.buttonValue[6] = jdata["buttons"][8].asDouble() > _button_threshold.get();
+        raw_cmd_obj.buttonValue[7] = jdata["buttons"][9].asDouble() > _button_threshold.get();
```
I thought we agreed we would not **hardcode** the mapping from web to rock here, but use configuration parameters.

doudou commented on test/web_socket_task_test.rb:
```diff
@@ -0,0 +1,214 @@
+            state.received_messages.append(message)
+        end
+    end
+    rescue Kontena::Websocket::CloseError => exc
```
`rescue` should have the same indent level than the method (i.e. aligned with `def`). Isn't this warned by rubocop ?

doudou commented on test/web_socket_task_test.rb:
```diff
@@ -0,0 +1,214 @@
+    run_live
+
+    def start_websocket (name)
+        $first_websockets[name] = WebsocketStruct.new(nil, [], :pending, false)
```
Leftover ? `$first_websockets` isn't used anywhere else.

doudou commented on test/web_socket_task_test.rb:
```diff
@@ -0,0 +1,214 @@
+    def start_websocket (name)
+        $first_websockets[name] = WebsocketStruct.new(nil, [], :pending, false)
+
+        thread = Thread.new {
```
The convention is to use `{ }` for single-line blocks and `do ... end` for multiline blocks. You've done this a few times, I won't be repeating it.

doudou reviewed (changes requested):
Looks good overall. Have you got rubocop working under vscode ? The ruby test file looks like it should have a lot of errors/warnings. I want those fixed before we merge the PR.

About the button/axis mapping: I'd be fine with merging this as-is while you fix it for the next pull request/version.

Commit:
Added types to help conversion and fixed style issues.

henriquejsfj reviewed (commented):
Still have to:
- [ ] Fix Rubocop issues to fix style issues in test code.
- [x] Update test code to verify new ports and set mapping properties.
- [x] Update CMakeLists to link Mapping.cpp with the header.

doudou commented on Mapping.hpp:
```diff
@@ -0,0 +1,39 @@
+         */
+        Type type;
+
+        std::string getType(){
```
`getXXX` should really only return the value from the corresponding field. The goal of this method is to return the JSON field that matches this mapping, so getJSONFieldName() would be a better name. In addition, a plain getter like this should be const.

Moreover, I think it would be better to not "leak" internal stuff like this in the interface type. Can add it as a static functions inside the task.cpp

henriquejsfj replied:
I don't agree with putting it in the task.cpp. The interface type is to do this. Ideally, if I change the browser side I should just update Mapping (except if I change status). For status maybe it is good to put it as a parameter, although it should not be changing it will be good as doc.

doudou replied:
This method is not an intrinsic operation of the mapping type. It is a function that makes the link between the type and the representation of the data structure in JSON, which is an operation specific to the task. Hence, I don't think it should be here.

doudou commented on Mapping.hpp:
```diff
@@ -0,0 +1,39 @@
+            if (type == Type::Axis){
+                return "axes";
+            }
+            return "";
```
I think you should throw here

doudou commented on tasks/Task.cpp:
```diff
@@ -45,42 +48,40 @@ struct controldev_websocket::JoystickHandler : WebSocket::Handler {
     }
 };
 
+Json::Value jdata;
```
Global variables ... :-1:  :-1:  :-1:

doudou commented on tasks/Task.cpp:
```diff
@@ -45,42 +48,40 @@ struct controldev_websocket::JoystickHandler : WebSocket::Handler {
-        raw_cmd_obj.buttonValue[7] = jdata["buttons"][9].asDouble() > _button_threshold.get();
+        auto axis = _axis_map.get();
+        for (uint i = 0; i < axis.size(); ++i){
+            raw_cmd_obj.axisValue[i] = jdata[axis[i].getType()][axis[i].index].asDouble();
```
Use `.at()` instead of `[]` to have bounds check.

doudou commented on tasks/Task.cpp:
```diff
@@ -45,42 +48,40 @@ struct controldev_websocket::JoystickHandler : WebSocket::Handler {
+        auto button = _button_map.get();
+        for (uint i = 0; i < button.size(); ++i){
+            raw_cmd_obj.buttonValue[i] = jdata[button[i].getType()][button[i].index]
+                                             .asDouble() > _button_threshold.get();
```
I think the threshold should be per-mapping

doudou commented on tasks/Task.cpp:
```diff
@@ -45,42 +48,40 @@ struct controldev_websocket::JoystickHandler : WebSocket::Handler {
-    if (!(jdata.isMember("status") && jdata.isMember("axes") && jdata.isMember("buttons"))) {
+    if (!jdata.isMember("status") || !jdata.isMember("axes") || !jdata.isMember("buttons")) {
+        LOG_ERROR_S << "Invalid message, it does not contain required fields." << std::endl;
+        ++errors;
```
Do the statistics 9both received and errors) in the caller of `updateRawCommand`, which will guarantee that we won't forget one error path.

doudou commented on tasks/Task.cpp:
```diff
@@ -126,17 +125,16 @@ bool Task::startHook()
 void Task::updateHook()
 {
     TaskBase::updateHook();
+    _received.write(received);
```
Make one __timestamped__ statistics structure and output it. Also, would IMO be better to output it in the callback for the handler (updateHook will generally not be called in this component)

henriquejsfj replied:
> (updateHook will generally not be called in this component)

Why not? I put the task to be periodic. Being periodic the updateHook is called for each period don't? Also, The statics being periodic is good to know if doesn't have data due to errors or is not receiving messages.

>  structure and output it

So, some code will need to read it and could be another component. Thus, isn't it better to keep all these structures in a Rock Library?

doudou replied:
> So, some code will need to read it and could be another component. Thus, isn't it better to keep all these structures in a Rock Library?

All these structures are specific to this particular component. So, using them will be useful only in relation to this particular component and can depend on this package.

In practice, the stats will be read mostly during log postprocessing or by Syskit.

*If* the case arises that the type is more generic that we think, we'll move it to a common package the same way we did for `gps_base` or `ais_base`

doudou replied:
> Why not? I put the task to be periodic. Being periodic the updateHook is called for each period don't? Also, The statics being periodic is good to know if doesn't have data due to errors or is not receiving messages.

What "errors" ?

Commit:
Update mapping structure and fixed issues.

Commit:
Update test with controldev_websocket. Still failing to set task properties.

henriquejsfj commented:
Just did the review, still need to fix the points at the top of the discussion.

doudou commented on controldev_websocket.orogen:
```diff
@@ -1,15 +1,23 @@
-task_context "Task" do
+task_context 'Task' do
     property 'port', 'uint16_t', 58000
+    property 'axis_map', '/std/vector</controldev_websocket/Mapping'
```
Does this work ? It's missing a `>` at the end ... If it does, it's a bug :P

henriquejsfj replied:
It is. At controldev.orogen is also the same way and I have to check again if the correct way works because I remeber I chaged it (maybe the change wasn't effective).

henriquejsfj replied:
It worked in the correct way

doudou commented on tasks/Task.cpp:
```diff
@@ -0,0 +1,167 @@
+bool Task::updateRawCommand(const char *data){
+    std::string errs;
+    // If an error occurs trying to get the parameters should not ignore it.
+    auto axis = _axis_map.get();
+    auto button = _button_map.get();
```
Accesssing properties at runtime is a Bad Idea

Properties are not thread-safe, and should be accessed only outside the runtime state. Make a copy of the two arrays internally.

doudou commented on tasks/Task.cpp:
```diff
@@ -0,0 +1,167 @@
+        return false;
+    }
+    if (!wrapper->jdata.isMember("status") ||
+        !wrapper->jdata.isMember(Mapping::mapFieldName(Mapping::Type::Axis)) ||
+        !wrapper->jdata.isMember(Mapping::mapFieldName(Mapping::Type::Button))) {
```
This is too much. Just put the strings directly.

henriquejsfj replied:
But is safer. Really prefer that?

doudou replied:
I see your point. I would feel a lot less weird if you would move mapFieldName to this file as I requested. It really has nothing to do anywhere else.

`mapFieldName`'s only goal is to be compatible with whatever JSON is processed by this task. It must be defined here.

henriquejsfj replied:
I moved yesterday @doudou 

Now is part of WrapperJSON

doudou commented on tasks/Task.cpp:
```diff
@@ -0,0 +1,167 @@
+                                                             > button.at(i).threshold;
+        }
+
+    // An exception here means a failure while converting the Gamepad. Just return false.
```
Exceptions always mean "a failure". Detail which failures might cause the exception.

doudou commented on tasks/Task.cpp:
```diff
@@ -0,0 +1,167 @@
+    server = new Server (logger);
+    handler = std::make_shared<JoystickHandler>();
+    handler->task = this;
+
+
```
Extra empty line

doudou commented on test/web_socket_task_test.rb:
```diff
@@ -0,0 +1,254 @@
+                           ButtonMapping. new,
+                           ButtonMapping. new]
+        thresh = 0.5
+        task.button_map[0].index = 0
+        task.button_map[2].index = 1
+        task.button_map[3].index = 2
+        task.button_map[4].index = 3
+        task.button_map[5].index = 4
+        task.button_map[6].index = 5
+        task.button_map[7].index = 8
+        task.button_map[8].index = 9
+        task.button_map[9].index = 12
+        task.button_map[10].index = 13
+        task.button_map[11].index = 14
+        task.button_map[12].index = 15
```
You're not setting the `type` field. Also,
- should be `task.properties`
- can't change properties on-the-fly. Writing a property accessor is what causes an update, e.g. `task.properties.button_map = ` is what triggers an update.
- https://www.rock-robotics.org/rock-and-syskit/components/types_in_ruby.html

Commit:
Fixed typo at Mapping.

Commit:
Fixing properties initialization.

Commit:
Passed all the types to the correct file and added statitics structure.

Commit:
Update the test to handle task updates.

Commit:
Update orogen file to handle all updates.

henriquejsfj commented:
I still need to fix robocop issues. Should I test the statistics port?

doudou commented on test/web_socket_task_test.rb:
```diff
@@ -0,0 +1,247 @@
+                to { have_no_new_sample task.raw_command_port, at_least_during: 1.0 }
+        end
+
+        it 'doesn\'t write when not controlling' do
```
What is "not controlling" ???

doudou replied:
Ping @henriquejsfj

doudou reviewed (commented):
I'm very puzzled by the  'status' flag. What is it for ?

doudou commented:
> I still need to fix robocop issues. Should I test the statistics port?

Yes.

henriquejsfj commented:
> I'm very puzzled by the 'status' flag. What is it for ?

The status flag is the answer to your question "What is "not controlling" ???".
`status = False` means not controlling.

This was first introduced at the client as a double check (should never be sent a message with status false). But then I get this advantage to send a first message when connecting to "verify" if the message can be processed correctly. So, I send a message full of zeros with status false and if the server answers `True` (I'm good) and then passes to controlling state (in client-side) and keep alive sending messages.

If you don't like it, time to review client-side: https://github.com/tidewise/gui-seabots_ui/pull/7

doudou commented:
I don't like it. At all.

If you want a "connection check", make it a different message.

Commit:
Fixed statistics issues.

Commit:
Add statistic port test.

henriquejsfj commented:
It is not just a connection check, it is checking if the message is being "decoded" correctly. 

And this is more important now with the mapping object since it is a parameter (configuration) now.

Also, if I would send a different message I should put something indicating that is a connection check" since the callback is the same. Thus, `status = False`.

doudou commented:
> It is not just a connection check, it is checking if the message is being "decoded" correctly.

Then, still, make it a different message with the "test message" as payload.

For instance,

~~~
{ "handshake": { "test_command": COMMAND } }
~~~

Based on what you say, the other side should receive a single "handshake" *first* before any other messages, and never again. Having a different message (and making sure this is how the javascript code handles it) ensures that. A boolean flag is not meant to encode something like that.

doudou commented:
Ping ?

Commit:
Add examples folder with HTML page to stream a gamepad through websocket.

henriquejsfj commented:
> Then, still, make it a different message with the "test message" as payload.
> 

Please, see the client PR about status also.

I don't agree with that, even if I changed to what you said the `status` flag would continuously existing at the message. Because it acts as a double security layer verifying if the client is controlling or not. 

In addition, I don't see why to create a new test field. To handle this change this code would verify if it contains a "handshake" with a "test_command" and being different should call another function to process it (but has to process equal the actual one) or maybe should preprocess the test message to create a nontest command like to pass to the function to process it and send the same result back to the client. And to make it worst, at the client-side this would become a new sender (function or whatever) just to test it.

For me, this is just making things more complicated.

henriquejsfj commented:
Added examples as requested when this was at `drivers-orogen-controldev`

doudou commented:
The `status` message is *not* part of the control message, it's part of the protocol. It should be false in the one and only one first message the client sends to check if it can start controlling.

Having two widely different messages make sure that you don't mistake one message for the other ... i.e. that there are no codepaths where 'status' has not been checked.

Moreover, I really don't agree with the whole point of `status`. The semantics of it in the client library is to say "I want control" and wait for the result on the other side that says "OK". Then just make it so. If you want to embed a sample control device message in it as part of the "can I control ?" query, then it's up to you (I don't really understand it, but I don't have a problem with that).

But the message should say "I want to start control" and the component should handle it as such. Not "here's a control but disregard it because there's a status flag that is false, so its semantic is actually different and only I (client) know why I would ever send it with a status == false. But, by the way, it is super important because I rely on the answer to tell me whether I can start controlling or not".

> In addition, I don't see why to create a new test field. To handle this change this code would verify if it contains a "handshake" with a "test_command" and being different should call another function to process it (but has to process equal the actual one)

If you can't share the message *parsing* code, then refactor the code so that you can (pseudocode):

~~~ cpp
Task::processControlStartHandshake(json) {
   try {
      parseControlMessage(json["test_message"])
  }
  catch(...) {
     INVALID
  }

  DO OTHER CHECKS, for instance: terminate existing clients here.
  return valid
}

Task::processControlMessage(json) {
   _outport.write(parseControlMessage(json);
}
~~~

>  And to make it worst, at the client-side this would become a new sender (function or whatever) just to test it.

Yes. And it should **even** in the current implementation. Because right now there's a sendMessage function that actually has two completely different meanings based on some internal state variable. Hence the question I had (what is that sendMessage doing here ?).

This is waaaay worse than having two functions with well-defined semantics that someone can see in their names, **sharing most of their internals**.

~~~
sendControlStartHandshake()
sendControlMessage()
~~~

> For me, this is just making things more complicated.

It makes more API functions. It also make *what* these functions less optional, dependent on behind-the-scenes changes in the internal state. More code, *yet* less complicated.

Commit:
Update client to do a handshake.

Commit:
Update server to do a handshake.

Commit:
Update test to test the handshake.

Commit:
Add readme file.

Commit:
Fix readme.

doudou commented on tasks/Task.cpp:
```diff
@@ -23,24 +21,37 @@ struct controldev_websocket::JoystickHandler : WebSocket::Handler {
+            } else {
+                Json::Value test_msg;
+                if (task->getTestMessage(data, test_msg)){
+                    auto str = fast.write(test_msg);
```
This smells _so bad_.

Refactor the whole thing to have `updateRawCommand` get a Json::Value object. Then, pass it *already decoded* in both codepaths. Factor out common code instead of strong-arming the existing code into new uses.

Also, there are too many levels of ifs in this method. Need to refactor to make the flow easier to follow. For instance (no error handling included):

~~~ cpp
if (connection == socket) {
   task->handleIncomingWebsocketMessage(data)
}
...

/**
 * Here, I'm not really sure of the value of WrapperJSON. I have the impression that it is only there for the purpose of getValue ?
 * I'd say, move the builder and JSON value out of WrapperJSON, passing only a const reference to the JSON value in the constructor instead. And move them to Task
 */

bool Task::handleIncomingWebsocketMessage(char const* data) {
    if (!parseJSONMessage(data, &jdata)) {
       return false;
    }

    if (is_controlling) {
       validateControlMessage(jvalue);
       parseControlMessage(jvalue, raw_cmd_obj);
       _raw_command.write(raw_cmd_obj);
    }
    else {
       validateAskControlMessage(jvalue);
    }
}

bool Task::validateAskControlMessage(JSON::Value const& value) {
    if (!wrapper->jdata.isMember("test_message")) {
        return false;
    }
    return validateControlMessage(jdata["test_message"]);
}

void Task::updateRawCommand(JSON::Value const& value) {
   WrapperJSON wrapper(value); // just gives us easy-to-use access to the JSON fields.
}
~~~

henriquejsfj replied:
I got your point, but I think is not so simple. I should have to verify after `validateControlMessage(jvalue)` if the message is valid to continue and after `parseControlMessage(jvalue, raw_cmd_obj)` to see if it was parsed.

Also, when doing a handshake the `parseControlMessage` should be called.

I will try to do the closest and cleanly possible to it.

Commit:
Update message flow control to be simpler.

Now the message is processed by a handler at the task level which
parses and validate the received JSON message through the Message
Decoder. This decoder was previously named WrapperJSON but now
Message Decoder makes more sense. Finally, the handler calls the
updateRawCommand function to fill the Raw Command object.

Note, since the handler parse the message the JSON is stored at
the decoder.

Commit:
Update statitics test to handle Task changes.

Now the Task count received to the test message, so increased the
expected number of received.

henriquejsfj commented:
Now the message is processed by a handler at the task level which
parses and validate the received JSON message through the Message
Decoder. This decoder was previously named WrapperJSON but now
Message Decoder makes more sense. Finally, the handler calls the
updateRawCommand function to fill the Raw Command object.

Note, since the handler parses the message the JSON is stored at
the decoder.

doudou reviewed (commented):
This is getting somewhere.

Still had been pretty surprised by the side effect of 'validateAskControlMessage' though. This is, again, you trying to make two code paths be the same for the sake of number of lines of code, at the cost of the number of codepaths. Just see the effect of `is_controlling` on handleIncomingWebsocketMessage. There is basically two completely different methods hidden in plain sight (the only common part is the JSON decoding). If *the same* condition appears three times in a given method, it means it really should be a completely different method.

First, move the JSON decoding part into the websocket handler *without* referring to the Decoder class. The rest of the code should only be manipulating JSON objects. It makes no sense at all to have strings beyond the confines of the websocket.

Then, the websocket callback should be the one dispatching based on the task state [1] e.g.

~~~ cpp
if (task->isControlling()) {
   task->handleIncomingControlMessage(json)
else
   task->handleIncomingAskControlMessage(json)
~~~

Finally, since you need to *validate* the control message in the ask control case, refactor the JSON-to-rawcommand decoding method into a side-effect-free method that decodes and validates it.

~~~ cpp
bool MessageDecoder::decodeRawCommand(Json::Value const& v, controldev::RawCommand& cmd) {
}
~~~

you can then use this method in two different ways in the two different modes, instead of the rather ugly change of internal state (in updateRawCommand) in a codepath that should only check for command validity (ask control)

With all of this, the task methods will be changed into two completely different methods, removing the codepath complexity you had in the current implementation:

~~~ cpp
Task::handleIncomingControlMessage(Json::Value json) {
    if (!decoder->validateControlMessage(json)) {
       return false;
    }
    if (!decoder->parseRawCommand(message, raw_cmd_obj)) {
       return false;
   }
   _raw_command.write(raw_cmd_obj);
   return true;
}

Task::handleIncomingAskControlMessage(Json::Value json) {
    if (!decoder->validateAskControlMessage(json)) {
       return false;
    }
    controldev::RawCommand test_cmd;
    is_controlling = decoder->parseRawCommand(json["test_results"], test_cmd);
    return is_controlling;
}
~~~

doudou committed:
properly resolve seasocks

doudou committed:
remove `type_export_policy :all`

Commit:
Improved code to reduce the code paths.

Now the handleIncomingWebsocketMessage function calls the
handleAskControlMessage and handleControlMessage depending on the
is_controlling status. Also, if AskControl fails the connection is
closed.

Commit:
Fixed typo and minor verifications update.
